CDP4CommonView/Diagram/ViewModels/PortContainerDiagramContentItem.cs

```diff
@@ -0,0 +1,169 @@
+            this.RecalculatePortsPosition();
+        }
+
+        /// <summary>
+        /// Recalculate all Ports position then fires <see cref="IDiagramPortViewModel.WhenPositionIsUpdatedInvoke"/>
+        /// todo: refactor
+        /// </summary>
+        private void RecalculatePortsPosition()
+        {
+            var diagramItem = (this.Parent as DiagramItem);
+
+            var bottomSide = this.PortCollection.Where(p => p.PortContainerShapeSide == PortContainerShapeSide.Bottom).ToArray();
+            var portion = this.CalculatePortion(PortContainerShapeSide.Bottom);
+
+            for (var index = 0; index < bottomSide.Count(); index++)
+            {
+                var vector = portion * (index + 1) - 10;
+                bottomSide[index].Position = System.Windows.Point.Add(diagramItem.Position, new Vector(vector, diagramItem.ActualHeight - (10)));
+                bottomSide[index].WhenPositionIsUpdatedInvoke();
+            }
+
+            var leftSide = this.PortCollection.Where(p => p.PortContainerShapeSide == PortContainerShapeSide.Left).ToArray();
+            portion = this.CalculatePortion(PortContainerShapeSide.Left);
+
+            for (var index = 0; index < leftSide.Count(); index++)
+            {
+                var vector = portion * (index + 1) - 10;
+                leftSide[index].Position = System.Windows.Point.Add(diagramItem.Position, new Vector(0 - (10), vector));
+                leftSide[index].WhenPositionIsUpdatedInvoke();
+            }
+
+            var topSide = this.PortCollection.Where(p => p.PortContainerShapeSide == PortContainerShapeSide.Top).ToArray();
+            portion = this.CalculatePortion(PortContainerShapeSide.Top);
+
+            for (var index = 0; index < topSide.Count(); index++)
+            {
+                var vector = portion * (index + 1) - 10;
+                topSide[index].Position = System.Windows.Point.Add(diagramItem.Position, new Vector(vector, 0 - (10)));
+                topSide[index].WhenPositionIsUpdatedInvoke();
+            }
+
+            var rightSide = this.PortCollection.Where(p => p.PortContainerShapeSide == PortContainerShapeSide.Right).ToArray();
+            portion = this.CalculatePortion(PortContainerShapeSide.Right);
+
+            for (var index = 0; index < rightSide.Count(); index++)
+            {
+                var vector = portion * (index + 1) - 10;
+                rightSide[index].Position = System.Windows.Point.Add(diagramItem.Position, new Vector(diagramItem.ActualWidth - (10), vector));
+                rightSide[index].WhenPositionIsUpdatedInvoke();
+            }
+        }
+
+        /// <summary>
+        /// Calculate the next available side where a port can join
+        /// </summary>
+        /// <returns></returns>
+        private PortContainerShapeSide GetAvailableSide()
+        {
+            var split = ((double)this.PortCollection.Count(p => p.PortContainerShapeSide > PortContainerShapeSide.Undefined)) / 4;
```

Comment:
I think using modulus will do a better job and create better (readable) code:
return (PortContainerShapeSide)((this.PortCollection.Count(p => p.PortContainerShapeSide > PortContainerShapeSide.Undefined) % 4) * 25);